readme.md
# router

A miniaturized (and fast af) router for the edge, and everywhere else.

```js
import { Router, text, json } from "@johno/router";

const router = new Router();

router
  .get("/", (req) => text("Hello world!"))
  .get("/hello/:name", (req, env, ctx, routeInfo) => {
    json({ name: routeInfo.params.name });
  });

router.handle(request, env, ctx);
```

## Features

- Simple, declarative API
- Built for the edge
- Standards-based (Request/Response/URL)
- Zero dependencies
- Tiny footprint
- ESM-only
- Written in TypeScript

## Motivation

`@johno/router` is built for edge environments including Cloudflare Workers, Lambda, Netlify Functions, Next.js, etc.
What makes `@johno/router` interesting is what it doesn't do.

**`@johno/router` _does not_:**

- Monkey patch the global `Request` and `Response` objects
- Route match with regex, it uses a custom parser
- Implement a bespoke handler API
- Use middleware, we recommend using lib functions
- Minify or obfuscate code (readable and debuggable output)

#### Handler API

`@johno/router` uses a simple handler API that is compatible with the [Request][] and [Response][] objects. It _augments_
edge function handlers with an additional argument, `routeInfo`. This means you can drop in `@johno/router` to any serverless
function environment and you will work with their existing APIs.

## Installation

Use your favorite package manager to install `@johno/router`:

```sh
npm install @johno/router

# pnpm add @johno/router
# yarn add @johno/router
# bun add @johno/router
```

## Usage

If you've used a router before, `@johno/router` will feel familiar.
To define a set of routes you:

- Instantiate a router
- Add routes for given paths and HTTP methods
- Pass each route a _handler_ function
- Call `handle` with the request and its environment

```js
import { Router, text } from "@johno/router";

const router = new Router();

router.get("/", (req, res) => {
  text("Hello world!");
});

router.get("/hello/:name", (req, res, routeInfo) => {
  text(`Hello ${routeInfo.params.name}!`);
});

router.handle(request, env, ctx);
```

### Using with Cloudflare Workers

`@johno/router` is designed to work seamlessly with Cloudflare Workers.
The `handle` method accepts the `Request`, `Env`, and `Context` objects
which you can leverage in your handlers.

```js
import { Router, text, json } from "@johno/router";

const router = new Router();

export default {
  async fetch(
    request: Request,
    env: Env,
    ctx: ExecutionContext
  ): Promise<Response> {
    return router
      .post("/hello", () => text("Hello world!"))
      .post("/world", () => json({ hello: "world" }))
      .handle(request, env, ctx);
  },
};
```

## Code of Conduct

By participating in our community and interacting with this repository, you agree
to abide by our [code of conduct](code-of-conduct.md).

## Development

To learn about how to contribute to this project, please read the [contributing guide](.github/contributing.md).

## License

[MIT](license.md)

---

> Built by [johno](https://johno.com)

[Request]: https://developer.mozilla.org/en-US/docs/Web/API/Request
[Response]: https://developer.mozilla.org/en-US/docs/Web/API/Response


CHANGELOG.md
# @johno/router

## 0.0.3

### Patch Changes

- 6885944: Add missing docs links

## 0.0.2

### Patch Changes

- 7029c5f: Improve documentation


src/util.test.ts
import { test, expect } from "bun:test";

import { cleanURLPath, getRequestPath, getURLPath } from "./util";

test("cleanURLPath removes trailing slashes", () => {
  expect(cleanURLPath("/test/")).toEqual("/test");
});

test("cleanURLPath removes duplicate slashes", () => {
  expect(cleanURLPath("/test//foo")).toEqual("/test/foo");
});

test("cleanURLPath removes trailing slashes and duplicate slashes", () => {
  expect(cleanURLPath("/test//foo/")).toEqual("/test/foo");
});

test("cleanURLPath does not remove a single slash", () => {
  expect(cleanURLPath("/")).toEqual("/");
});

test("getRequestPath returns the pathname from a Request", () => {
  const request = new Request("https://example.com/test");

  expect(getRequestPath(request)).toEqual("/test");
});

test("getURLPath returns the pathname from a string", () => {
  expect(getURLPath("https://example.com/test")).toEqual("/test");
});

test("getURLPath returns the pathname from a string with a path", () => {
  expect(getURLPath("/test")).toEqual("/test");
});

test("getURLPath returns the pathname from a string with a path and query", () => {
  expect(getURLPath("/test?foo=bar")).toEqual("/test?foo=bar");
});

test("getURLPath returns the pathname from a string with a path and hash", () => {
  expect(getURLPath("/test#foo")).toEqual("/test#foo");
});

test("getURLPath returns the pathname from a string with a path, query, and hash", () => {
  expect(getURLPath("/test?foo=bar#baz")).toEqual("/test?foo=bar#baz");
});


src/parse.test.ts
import { test, expect } from "bun:test";

import { parse } from "./parse";

test("parses a static route", () => {
  const result = parse("/test");

  expect(result).toEqual({
    raw: "/test",
    parsed: [{ type: "static", value: "test" }],
    matchParams: expect.any(Function),
    isMatch: expect.any(Function),
  });

  expect(result.matchParams("/test")).toEqual({});
  expect(result.isMatch("/test")).toEqual(true);
  expect(result.isMatch("/test/bar")).toEqual(false);
});

test("parses a route with a named parameter", () => {
  const result = parse("/foo/:bar");

  expect(result).toEqual({
    raw: "/foo/:bar",
    parsed: [
      { type: "static", value: "foo" },
      { type: "parameter", value: "bar" },
    ],
    matchParams: expect.any(Function),
    isMatch: expect.any(Function),
  });

  expect(result.matchParams("/foo/test")).toEqual({ bar: "test" });
  expect(result.isMatch("/foo/test")).toEqual(true);
  expect(result.isMatch("/foo")).toEqual(false);
});

test("parses a route with a catch all", () => {
  const result = parse("/foo/*");

  expect(result).toEqual({
    raw: "/foo/*",
    parsed: [
      { type: "static", value: "foo" },
      { type: "catchAll", value: "*" },
    ],
    matchParams: expect.any(Function),
    isMatch: expect.any(Function),
  });

  expect(result.matchParams("/foo/test")).toEqual({});
  expect(result.isMatch("/foo/test/bar")).toEqual(true);
  expect(result.isMatch("/foo")).toEqual(true);
});

test("parses a route with a named parameter and ignores trailing slashes", () => {
  const result = parse("/foo/:bar");

  expect(result).toEqual({
    raw: "/foo/:bar",
    parsed: [
      { type: "static", value: "foo" },
      { type: "parameter", value: "bar" },
    ],
    matchParams: expect.any(Function),
    isMatch: expect.any(Function),
  });

  expect(result.matchParams("/foo/test/")).toEqual({ bar: "test" });
  expect(result.isMatch("/foo/test/")).toEqual(true);
  expect(result.isMatch("/foo")).toEqual(false);
});


src/router.test.ts
import { test, expect, beforeEach } from "bun:test";

import { Router, json } from "./router";

let router: Router;

beforeEach(() => {
  router = new Router();
});

test("routes to a path", async () => {
  router.get("/test", async () => new Response("test"));

  const response = await router.handle(
    new Request("https://example.com/test"),
    {} as any,
    {} as any
  );

  expect(await response.text()).toEqual("test");
});

test("returns a 404 for a missing path", async () => {
  const response = await router.handle(
    new Request("https://example.com/missing"),
    {} as any,
    {} as any
  );

  expect(response.status).toEqual(404);
});

test("returns a 405 for a missing method", async () => {
  router.get("/test", async () => new Response("test"));

  const response = await router.handle(
    new Request("https://example.com/test", { method: "POST" }),
    {} as any,
    {} as any
  );

  expect(response.status).toEqual(405);
});

test("returns a 500 for an error", async () => {
  router.get("/test", async () => {
    throw new Error("test");
  });

  const response = await router.handle(
    new Request("https://example.com/test"),
    {} as any,
    {} as any
  );

  expect(response.status).toEqual(500);
});

test("sets headers and stringifies a JSON response", async () => {
  router.get("/test", async () => json({ test: true }));

  const response = await router.handle(
    new Request("https://example.com/test"),
    {} as any,
    {} as any
  );

  const result = await response.json();

  expect(response.headers.get("content-type")).toEqual("application/json");
  expect(result).toEqual({ test: true });
});

test("handles CORS preflight requests", async () => {
  router.get("/test", async () => new Response("test"));

  const response = await router.handle(
    new Request("https://example.com/test", {
      method: "OPTIONS",
      headers: {
        Origin: "https://example.com",
        "Access-Control-Request-Method": "GET",
        "Access-Control-Request-Headers": "Content-Type",
      },
    }),
    {} as any,
    {} as any
  );

  expect(response.status).toEqual(204);
  expect(response.headers.get("access-control-allow-origin")).toEqual("*");
  expect(response.headers.get("access-control-allow-methods")).toEqual(
    "GET, POST, DELETE"
  );
  expect(response.headers.get("access-control-allow-headers")).toEqual(
    "Content-Type"
  );
});

test("matches HTTP methods", async () => {
  router.get("/test", async () => new Response("test GET"));
  router.post("/test", async () => new Response("test POST"));
  router.delete("/test", async () => new Response("test DELETE"));

  const getResponse = await router.handle(
    new Request("https://example.com/test"),
    {} as any,
    {} as any
  );
  const postResponse = await router.handle(
    new Request("https://example.com/test", { method: "POST" }),
    {} as any,
    {} as any
  );
  const deleteResponse = await router.handle(
    new Request("https://example.com/test", { method: "DELETE" }),
    {} as any,
    {} as any
  );

  expect(await getResponse.text()).toEqual("test GET");
  expect(await postResponse.text()).toEqual("test POST");
  expect(await deleteResponse.text()).toEqual("test DELETE");
});

test("passes route params to the handler", async () => {
  router.get("/test/:id", async (_request, _env, _ctx, routeInfo) =>
    json(routeInfo?.params ?? {})
  );

  const response = await router.handle(
    new Request("https://example.com/test/123"),
    {} as any,
    {} as any
  );

  const result = await response.json();

  expect(result).toEqual({ id: "123" });
});

test("ignores duplicate and trailing slashes", async () => {
  router.get("/test///foo/", async (_request, _env, _ctx) =>
    json({ hello: "world" })
  );

  const response = await router.handle(
    new Request("https://example.com/test/foo"),
    {} as any,
    {} as any
  );

  const result = await response.json();

  expect(result).toEqual({ hello: "world" });
});


src/parse.ts
import {
  CATCH_ALL_TYPE,
  PARAMETER_TYPE,
  STATIC_TYPE,
  URL_SLASH,
} from "./constants";
import { tokenize } from "./tokenize";
import { RouteParams, ParsedRoute } from "./types";
import { cleanURLPath, getURLPath } from "./util";

export function parse(route: string): ParsedRoute {
  const routeTokens = tokenize(route);

  const matchParams = (url: string) => {
    const urlParts = cleanURLPath(url).split(URL_SLASH).filter(Boolean);
    let params: RouteParams = {};

    if (
      routeTokens.length !== urlParts.length &&
      !routeTokens.some((token) => token.type === CATCH_ALL_TYPE)
    ) {
      return null;
    }

    for (let i = 0; i < routeTokens.length; i++) {
      const token = routeTokens[i];

      if (token.type === CATCH_ALL_TYPE) {
        return params;
      }

      if (token.type === PARAMETER_TYPE) {
        params[token.value] = urlParts[i];
      } else if (token.type === STATIC_TYPE && token.value !== urlParts[i]) {
        return null;
      }
    }

    return params;
  };

  function isMatch(providedUrl: string) {
    const url = cleanURLPath(providedUrl);

    if (matchParams(url)) {
      return true;
    }

    return getURLPath(url) === route;
  }

  return {
    raw: route,
    parsed: routeTokens,
    matchParams: matchParams,
    isMatch,
  };
}


src/types.ts
export type HandledResponse = Promise<Response>;
export type HTTPMethod = "GET" | "POST" | "OPTIONS" | "HEAD" | "PUT" | "DELETE";
export type Env = any;
export type ExecutionContext = any;
export type RouteInfo = {
  params: Record<string, string>;
  path: string;
  query: URLSearchParams;
};
export type RouteHandler = (
  request: Request,
  env: Env,
  ctx: ExecutionContext,
  routeInfo?: RouteInfo
) => HandledResponse;
export type RouteParams = { [key: string]: string };
export type RouteToken = {
  type: "static" | "parameter" | "catchAll";
  value: string;
};
export type ParsedRoute = {
  raw: string;
  parsed: RouteToken[];
  matchParams: (url: string) => RouteParams | null;
  isMatch: (url: string) => boolean;
};
type JSONResponseObjectValuePrimitive = string | number | boolean | null;
type JSONResponseObjectValue =
  | JSONResponseObjectValuePrimitive
  | JSONResponseObjectValuePrimitive[]
  | Record<string, JSONResponseObjectValuePrimitive>;
export type JSONResponseObject = Record<string, JSONResponseObjectValue>;


src/router.ts
import {
  DEFAULT_PARAMS,
  corsHeaders,
  mimeTypes,
  statusCodes,
} from "./constants";
import { parse } from "./parse";
import {
  Env,
  ExecutionContext,
  HTTPMethod,
  JSONResponseObject,
  ParsedRoute,
  RouteHandler,
  RouteInfo,
} from "./types";
import { getRequestPath } from "./util";

export function json(obj: JSONResponseObject) {
  return new Response(JSON.stringify(obj) + "\n", {
    status: statusCodes.ok,
    headers: {
      "Content-Type": mimeTypes.json,
    },
  });
}

export function text(str: string) {
  return new Response(str, {
    status: statusCodes.ok,
    headers: {
      "Content-Type": mimeTypes.plain,
    },
  });
}

export function error(status: number, message?: string) {
  return new Response(
    JSON.stringify({
      error: true,
      message: message ?? "An error occurred",
    }) + "\n",
    { status }
  );
}

class Route {
  constructor(
    private method: HTTPMethod,
    private route: ParsedRoute,
    private handler: RouteHandler
  ) {}

  matches(request: Request): boolean {
    return request.method === this.method && this.hasMatchingPath(request);
  }

  hasMatchingPath(request: Request): boolean {
    return this.route.isMatch(getRequestPath(request));
  }

  routeInfo(request: Request): RouteInfo {
    const url = new URL(request.url);

    return {
      params: this.route.matchParams(url.pathname) || DEFAULT_PARAMS,
      path: url.pathname,
      query: url.searchParams,
    };
  }

  async execute(request: Request, env: Env, ctx: ExecutionContext) {
    const routeInfo = this.routeInfo(request);
    return await this.handler(request, env, ctx, routeInfo);
  }
}

export class Router {
  private routes: Route[] = [];

  get(path: string, handler: RouteHandler) {
    this.addRoute("GET", path, handler);
    return this;
  }

  post(path: string, handler: RouteHandler) {
    this.addRoute("POST", path, handler);
    return this;
  }

  put(path: string, handler: RouteHandler) {
    this.addRoute("PUT", path, handler);
    return this;
  }

  delete(path: string, handler: RouteHandler) {
    this.addRoute("DELETE", path, handler);
    return this;
  }

  head(path: string, handler: RouteHandler) {
    this.addRoute("HEAD", path, handler);
    return this;
  }

  async handle(request: Request, env: Env, ctx: ExecutionContext) {
    let hasMissingMethod = false;

    const corsResponse = this.handleCORS(request);
    if (corsResponse.status !== 200) {
      return corsResponse;
    }

    try {
      for (const route of this.routes) {
        if (route.matches(request)) {
          return await route.execute(request, env, ctx);
        } else if (route.hasMatchingPath(request)) {
          hasMissingMethod = true;
        }
      }
    } catch (e) {
      return error(statusCodes.internalServerError, "Internal server error");
    }

    if (hasMissingMethod) {
      return error(statusCodes.methodNotAllowed, "Method not allowed");
    }

    return error(statusCodes.notFound, "Not found");
  }

  private addRoute(method: HTTPMethod, path: string, handler: RouteHandler) {
    const parsedRoute = parse(path);
    this.routes.push(new Route(method, parsedRoute, handler));
    return this;
  }

  private handleCORS(request: Request): Response {
    const origin = request.headers.get("Origin");

    if (origin !== null) {
      if (request.method === "OPTIONS") {
        return new Response(null, {
          status: statusCodes.noContent,
          headers: corsHeaders,
        });
      }

      return new Response(null, {
        headers: { ...corsHeaders, "Content-Type": mimeTypes.plain },
        status: statusCodes.noContent,
      });
    }

    return new Response(null);
  }
}


src/constants.ts
export const corsHeaders = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "GET, POST, DELETE",
  "Access-Control-Allow-Headers": "Content-Type",
};

export const mimeTypes = {
  plain: "text/plain",
  json: "application/json",
  html: "text/html",
  css: "text/css",
  js: "application/javascript",
  png: "image/png",
  jpg: "image/jpeg",
  gif: "image/gif",
  svg: "image/svg+xml",
  mp4: "video/mp4",
  webm: "video/webm",
  webp: "image/webp",
  ico: "image/x-icon",
  woff: "font/woff",
  woff2: "font/woff2",
  ttf: "font/ttf",
  otf: "font/otf",
  xml: "application/xml",
  pdf: "application/pdf",
  zip: "application/zip",
};

export const statusCodes = {
  ok: 200,
  noContent: 204,
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  methodNotAllowed: 405,
  internalServerError: 500,
};

export const URL_SLASH = "/";
export const NAMED_PARAM_TOKEN = ":";
export const CATCH_ALL_TOKEN = "*";

export const STATIC_TYPE = "static";
export const PARAMETER_TYPE = "parameter";
export const CATCH_ALL_TYPE = "catchAll";

export const DEFAULT_PARAMS = {};
